File: metrics.go
package main

import (
	"sort"
)

type Metrics struct {
	Throughput float64
	Delay      float64
	AoI        float64
}

// GetOneIndex efficiently finds the first successful transmission in each interval of size T.
// sIndex must be sorted.
func GetOneIndex(sIndex []int, T int) []int {
	if len(sIndex) == 0 {
		return nil
	}
	var oneIndex []int
	lastGroup := -1
	for _, val := range sIndex {
		group := val / T
		if group > lastGroup {
			oneIndex = append(oneIndex, val)
			lastGroup = group
		}
	}
	return oneIndex
}

// CalculateMetrics calculates Throughput, Delay, and AoI for a single user's success indices.
func CalculateMetrics(successIndices []int, T, pq int) Metrics {
	throughput := float64(len(successIndices))

	if len(successIndices) == 0 {
		// If no successful transmissions, return throughput = 0, delay = 0, AoI = 0
		// (Matching MATLAB logic where delayCount=0 skips delay addition)
		return Metrics{0, 0, 0}
	}

	// 1. Calculate Delay
	delaySum := 0.0
	delayCount := 0

	// We iterate through groups from g=1 to pq/T
	numGroups := pq / T
	lastGroup := -1

	// Since successIndices is sorted, we can find the first element of each group
	for _, val := range successIndices {
		group := val / T
		if group < numGroups && group > lastGroup {
			modValue := val % T
			if modValue == 0 {
				delaySum += 1.0
			} else {
				delaySum += float64(modValue + 1)
			}
			delayCount++
			lastGroup = group
		}
	}

	delayAvg := 0.0
	if delayCount > 0 {
		delayAvg = delaySum / float64(delayCount)
	}

	// 2. Calculate AoI
	oneIndex := GetOneIndex(successIndices, T)

	// Create extendedMap array for O(1) lookup
	// Using []bool is dramatically faster than map[int]bool for dense integer ranges
	extendedMap := make([]bool, 2*pq+1) // +1 to safely access indices up to 2*pq
	for _, val := range oneIndex {
		if val < len(extendedMap) {
			extendedMap[val] = true
		}
		if val+pq < len(extendedMap) {
			extendedMap[val+pq] = true // second_group
		}
	}

	extendedAoI := make([]float64, 2*pq)
	for d := 1; d <= 2*pq; d++ {
		// To match MATLAB exactly: it checks ismember(d, extended_Index) where d starts at 1
		if extendedMap[d] {
			extendedAoI[d-1] = float64((d % T) + 1)
		} else if d == 1 {
			extendedAoI[d-1] = 0
		} else {
			extendedAoI[d-1] = extendedAoI[d-2] + 1
		}
	}

	// Calculate sum of the second half
	aoiSum := 0.0
	for d := pq; d < 2*pq; d++ {
		aoiSum += extendedAoI[d]
	}
	aoiAvg := aoiSum / float64(pq)

	return Metrics{Throughput: throughput, Delay: delayAvg, AoI: aoiAvg}
}

// CalculateSystemMetrics processes all users in one run.
// It mutates successCounts to ensure it's sorted, then calculates metrics.
func CalculateSystemMetrics(successCounts [][]int, T, pq int) []Metrics {
	se := len(successCounts)
	res := make([]Metrics, se)
	for i := 0; i < se; i++ {
		// Must sort because ApplyRandomShifts disrupts the order, and MATLAB's find() is sorted.
		sort.Ints(successCounts[i])
		res[i] = CalculateMetrics(successCounts[i], T, pq)
	}
	return res
}

File: python_version/metrics.py
import numpy as np

def calculate_metrics(success_indices, T, pq):
    throughput = len(success_indices)
    
    if throughput == 0:
        return 0.0, 0.0, 0.0
        
    # 1. Delay Calculation
    # Find the first successful index in each interval of size T
    group_ids = success_indices // T
    _, first_idxs = np.unique(group_ids, return_index=True)
    first_vals = success_indices[first_idxs]
    
    mod_values = first_vals % T
    delays = np.where(mod_values == 0, 1, mod_values + 1)
    delay_avg = np.mean(delays) if len(delays) > 0 else 0.0
    
    # 2. AoI Calculation
    # We must match the MATLAB/Go logic exactly.
    # extended_indices (0-indexed)
    extended_indices = np.concatenate([first_vals, first_vals + pq])
    
    # Create a boolean map for O(1) vectorized lookups
    # Size 2*pq + 1 to safely access up to 2*pq
    extended_map = np.zeros(2 * pq + 1, dtype=bool)
    # In Python, indices out of bounds will throw, but extended_indices <= 2pq-1
    extended_map[extended_indices] = True
    
    # d goes from 1 to 2pq
    d = np.arange(1, 2 * pq + 1)
    
    # Check if 'd' is in extended_map
    valid_mask = extended_map[d]
    
    # We want to fill an array aoi of size 2*pq.
    base_values = np.zeros(2 * pq, dtype=np.float64)
    base_values[valid_mask] = (d[valid_mask] % T) + 1
    
    # To vectorize the loop: 
    # if valid_mask[i]: aoi[i] = base_values[i]
    # elif i == 0: aoi[i] = 0
    # else: aoi[i] = aoi[i-1] + 1
    
    # We can use the fill-forward technique with np.maximum.accumulate
    # We mark the reset points. i=0 is a reset point unconditionally (if not valid_mask).
    reset_points = valid_mask.copy()
    reset_points[0] = True
    
    # Indices of reset points
    reset_indices = np.where(reset_points)[0]
    
    # For each position, find the index of the last reset point
    idx = np.searchsorted(reset_indices, np.arange(2 * pq), side='right') - 1
    last_reset_pos = reset_indices[idx]
    
    # Calculate AoI = Base value at last reset + distance from last reset
    aoi = base_values[last_reset_pos] + (np.arange(2 * pq) - last_reset_pos)
    
    # Calculate average of the second half
    aoi_avg = np.mean(aoi[pq:])
    
    return float(throughput), float(delay_avg), float(aoi_avg)

def calculate_system_metrics(success_counts, T, pq):
    """
    Processes all users in one run.
    """
    se = len(success_counts)
    throughputs = np.zeros(se, dtype=np.float64)
    delays = np.zeros(se, dtype=np.float64)
    aois = np.zeros(se, dtype=np.float64)
    
    for i in range(se):
        th, de, aoi = calculate_metrics(success_counts[i], T, pq)
        throughputs[i] = th
        delays[i] = de
        aois[i] = aoi
        
    return throughputs, delays, aois
